firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/Chunks.cpp

```diff
@@ -0,0 +1,427 @@
+#include "Pufferfish/Util/Vector.h"
+#include "catch2/catch.hpp"
+
+namespace PF = Pufferfish;
+
+SCENARIO(
+    "Protocols::ChunkSplitter behaves correctly across chunk boundary(102 bytes)", "[chunks]") {
+  GIVEN("A char chunk splitter with internal buffer equal to 102 bytes") {
+    constexpr size_t buffer_size = 102;
+    PF::Protocols::ChunkSplitter<buffer_size, char> chunks;
+
+    WHEN("the input data exceeds bounds after output is called, without delimiter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus input_status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        input_status = chunks.input(val);
+        THEN("the input status is equal to ok") {
+          REQUIRE(input_status == PF::Protocols::ChunkInputStatus::ok);
+        }
+      }
+
+      PF::Util::Vector<char, buffer_size> buffer;
+      PF::Protocols::ChunkOutputStatus output_status = chunks.output(buffer);
+      THEN("the output status is equal to waiting") {
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::waiting);
+      }
+      THEN("the output buffer is empty") {
+        REQUIRE(buffer.empty() == true);
+      }
+
+      auto final_status = chunks.input(val);
+      THEN("the final status is equal to invalid_length") {
+        REQUIRE(final_status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+  }
+}
+
+SCENARIO(
+    "Protocols::ChunkSplitter behaves correctly across chunk boundary(256 bytes)", "[chunks]") {
+  GIVEN("A uint8_t chunk splitter of internal buffer equal to 256 bytes") {
+    constexpr size_t buffer_size = 256;
+    PF::Protocols::ChunkSplitter<buffer_size, uint8_t> chunks;
+
+    WHEN("enough data is input to exactly fill the buffer of the splitter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the final status is equal to ok") {
+          REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        }
+      }
+    }
+
+    WHEN("the input data exceeds bounds without output being called") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the initial status is equal to ok") {
+          REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        }
+      }
+
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the final status is equal to invalid_length") {
+          REQUIRE(status == PF::Protocols::ChunkInputStatus::invalid_length);
+        }
+      }
+
+    }
+
+    WHEN("the input data exceeds bounds after output is called, without delimiter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+        THEN("the initial status is equal to ok") {
```

Comment:
This THEN doesn't belong in this WHEN, I think this `for` loop and the `REQUIRE` inside belongs under a Given a uint8_t chunk splitter with a full 256-byte internal buffer, outside any WHEN. Since it's an assertion on the state of the object as it's being set up.

Reply:
Moved.